### PR TITLE
Feature/admission controller

### PR DESCRIPTION
**Pull request by amitschendel**
## Overview
This PR fixes #

**[Signed Commits](../CONTRIBUTING.md#sign-off-per-commit)**
- [x] Yes, I signed my commits.

### CODE REVIEW

File: admission/webhook/server.go [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+				oldObject = &oldUnstructured
+			case err != nil:
+				failure(err, http.StatusBadRequest)
+				return
+			default:
+				oldObject = obj
+			}
+		}
+
+		if len(parsed.Request.Object.Raw) > 0 {
```
Comment (matthyx) [inline admission/webhook/server.go:250]: can we DRY it a bit?

File: admission/webhook/validator.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+	}
+}
+
+// We are implementing the Handles method from the ValidationInterface interface.
+// This method returns true if this admission controller can handle the given operation, we accept all operations.
+func (av *AdmissionValidator) Handles(operation admission.Operation) bool {
+	return true
+}
+
+func (av *AdmissionValidator) validateAdminRoleBinding(attrs admission.Attributes) error {
```
Comment (matthyx) [inline admission/webhook/validator.go:61]: I think at some point these will reside as separate rules, such as in node-agent?

**Review by matthyx — CHANGES_REQUESTED** [review]
interesting... some clarifications needed, but overall good
did you follow any example for the admission controller that you could document?

File: admission/webhook/validator.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+		}
+	}
+
+	return nil
+}
+
+func (av *AdmissionValidator) validatePods(attrs admission.Attributes) error {
+	var errs error
+
+	// Check if the request is for pod exec or attach.
```
Comment (dwertent) [inline admission/webhook/validator.go:119]: This is not "pod validation". You should have a different function for this

File: admission/webhook/validator.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+	if attrs.GetSubresource() == "exec" {
+		av.exporter.SendAdmissionAlert(&attrs, "R2000", fmt.Sprintf("exec to pod %s", attrs.GetName()))
+		errs = errors.Join(errs, admission.NewForbidden(attrs, fmt.Errorf("exec to pod is audited")))
+	}
+
+	if attrs.GetSubresource() == "attach" {
```
Comment (dwertent) [inline admission/webhook/validator.go:125]: Same, not pod validation

File: admission/webhook/validator.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+	}
+
+	// Check if the request is for pod with insecure capabilities (SYS_ADMIN, SYS_MODULE, NET_ADMIN, NET_RAW, SYS_PTRACE, SYS_BOOT, SYS_RAWIO, BPF).
+	if attrs.GetOperation() == admission.Create {
+		pod, ok := attrs.GetObject().(*v1.Pod)
+		if !ok {
+			return nil
+		}
+
+		for _, container := range pod.Spec.Containers {
```
Comment (dwertent) [inline admission/webhook/validator.go]: What about other containers?

File: admission/webhook/validator.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+			// If a volume is a hostPath, return an error.
+			if volume.HostPath != nil {
+				av.exporter.SendAdmissionAlert(&attrs, "R2004", fmt.Sprintf("hostPath volume in pod %s", attrs.GetName()))
+				errs = errors.Join(errs, admission.NewForbidden(attrs, fmt.Errorf("hostPath volume is audited")))
+			}
+		}
+	}
+
+	// Check if the request is for port-forwarding.
+	if attrs.GetSubresource() == "portforward" {
```
Comment (dwertent) [inline admission/webhook/validator.go:131]: Same, not pod validation

File: main.go [hunk 7]
```diff
@@ -136,6 +142,36 @@ func main() {
 		}(mainHandler)
 	}
 
+	if operatorConfig.AdmissionControllerEnabled() {
+		// Handle SIGINT and SIGTERM by cancelling the root context
+		admissionContext, stop := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
```
[the remark this answers is not in the document]
Comment (amitschendel) [reply main.go]: done

File: admission/webhook/validator.go (hunk 2 above)
Comment (amitschendel) [reply admission/webhook/validator.go:60]: Maybe at some point but no now.

File: admission/exporter/http_exporter.go [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+		return
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		logger.L().Error("Received non-2xx status code", helpers.Int("status", resp.StatusCode))
+		return
+	}
+
+	// discard the body
```
[the remark this answers is not in the document]
Comment (amitschendel) [reply admission/exporter/http_exporter.go:185]: I prefer to do it anyway to avoid potential leaks.

File: admission/webhook/server.go (hunk 1 above)
Comment (amitschendel) [reply admission/webhook/server.go:250]: wdym?

File: admission/webhook/validator.go [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+	if attrs.GetOperation() == admission.Create {
+		pod, ok := attrs.GetObject().(*v1.Pod)
+		if !ok {
+			return nil
+		}
+
+		for _, container := range pod.Spec.Containers {
+			if container.SecurityContext != nil && container.SecurityContext.Capabilities != nil {
+				for _, capability := range container.SecurityContext.Capabilities.Add {
+					if capability == "SYS_ADMIN" || capability == "SYS_MODULE" || capability == "NET_ADMIN" || capability == "NET_RAW" || capability == "SYS_PTRACE" || capability == "SYS_BOOT" || capability == "SYS_RAWIO" || capability == "BPF" {
```
Comment (dwertent) [inline admission/webhook/validator.go]: Use a list or a different structure, not like this.

**Review by dwertent — DISMISSED** [review]
I added some comments. 

Here are my overall comments:
1. What are the 'rules'? Why are they not defined as rules, but rather as "if" statements in a large function?
2. How can the user interact with this system?
3. This code is missing unit tests.
4. Some conditions, such as capabilities, are hard-coded. Consider implementing a more flexible solution.

I don't know the full context, but if this is just a POC, I don't see a reason to merge it into the main branch. You should continue working on it, and once it is more complete (at least the basics), we can consider merging it into the main branch.

File: main.go [hunk 10]
```diff
@@ -136,6 +142,36 @@ func main() {
...
+		defer stop()
+
+		waitGroup := sync.WaitGroup{}
```
[the remark this answers is not in the document]
Comment (amitschendel) [reply main.go:147]: Removed it but I think it's good to have for scaling.

File: admission/webhook/server.go [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+		}
+	}()
+	return res
+}
+
+func (wh *webhook) Run(ctx context.Context) error {
+	var serverError error
+	var wg sync.WaitGroup
+
+	log.Info("starting webhook HTTP server")
```
[the remark this answers is not in the document]
Comment (amitschendel) [reply admission/webhook/server.go]: done

File: admission/webhook/server.go [hunk 12]
```diff
@@ -0,0 +1,402 @@
...
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/authentication/user"
+)
+
+type Interface interface {
```
[the remark this answers is not in the document]
Comment (amitschendel) [reply admission/webhook/server.go]: done

File: main.go [hunk 13]
```diff
@@ -136,6 +142,36 @@ func main() {
...
+		waitGroup.Add(1)
+		go func() {
+			defer func() {
+				// Cancel the server context to stop other workers
+				serverCancel()
+				waitGroup.Done()
+			}()
+
+			cancellationReason := admissionController.Run(serverContext)
+			logger.L().Ctx(ctx).Error("server stopped", helpers.Error(cancellationReason))
```
[the remark this answers is not in the document]
Comment (amitschendel) [reply main.go]: done

File: main.go [hunk 14]
```diff
@@ -136,6 +142,36 @@ func main() {
...
+		// Start HTTP REST server for webhook
+		waitGroup.Add(1)
+		go func() {
+			defer func() {
+				// Cancel the server context to stop other workers
+				serverCancel()
+				waitGroup.Done()
+			}()
+
+			cancellationReason := admissionController.Run(serverContext)
```
[the remark this answers is not in the document]
Comment (amitschendel) [reply main.go]: done

**Comment by dwertent** [comment]
@amitschendel 

> Do you think we should use the rules packages from the node-agent?

Not really, because here this is a different set of rules, but I will expect to see a similar code structure. Consider using the same rule IDs as the regolibrary, meaning the isPrivileged should have the same control ID as we have with compliance scanning. This should be considered, I'm not sure we want it.

> Which kind of interaction do you expect? rule bindings?

Possibly. What were the requirements and design for the admission controller?

File: main.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (dwertent) [reply main.go]: I'm not sure what you mean by the "scaling", what is the scenario?

File: admission/exporter/http_exporter.go [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+func (exporter *HTTPExporter) sendInAlertList(httpAlert apitypes.RuntimeAlert, processTree apitypes.ProcessTree) {
+	// create the HTTPAlertsListSpec struct
+	// TODO: accumulate alerts and send them in a batch
+	httpAlertsListSpec := HTTPAlertsListSpec{
+		Alerts:      []apitypes.RuntimeAlert{httpAlert},
+		ProcessTree: processTree,
+	}
+	// create the HTTPAlertsList struct
+	httpAlertsList := HTTPAlertsList{
+		Kind:       "RuntimeAlerts",
```
[the remark this answers is not in the document]
Comment (dwertent) [reply admission/exporter/http_exporter.go:151]: This is the same kind of alert as coming from the node-agent, in other words, we will not know "down the line" the source of the alert. And I think we will want to know...

File: admission/exporter/http_exporter.go [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+	}
+
+	logger.L().Error("Alert limit reached", helpers.Int("alerts", exporter.alertCount), helpers.String("since", exporter.alertCountStart.Format(time.RFC3339)))
+	exporter.sendInAlertList(httpAlert, apitypes.ProcessTree{})
+}
+
+func (exporter *HTTPExporter) SendAdmissionAlert(admissionAttrs *admission.Attributes, ruleID string, RuleDescription string) {
+	isLimitReached := exporter.checkAlertLimit()
+	if isLimitReached {
+		exporter.sendAlertLimitReached()
```
[the remark this answers is not in the document]
Comment (amitschendel) [reply admission/exporter/http_exporter.go:114]: done

File: admission/exporter/http_exporter.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (amitschendel) [reply admission/exporter/http_exporter.go:149]: done

**Review by matthyx — CHANGES_REQUESTED** [review]
1. fix the conflicts with go.mod
2. remove github.com/zeebo/assert and use testify instead (as everywhere)
3. improve some of the test coverage (for example the cache)
4. I wonder if we should have a component test or something testing the full functionality of the webhook (outside of the system tests that you will add)

File: admission/rules/v1/r2001_portforward_test.go [hunk 17]
```diff
@@ -0,0 +1,46 @@
+package rules
+
+import (
+	"testing"
+
+	"github.com/zeebo/assert"
```
[the remark this answers is not in the document]
Comment (amitschendel) [reply admission/rules/v1/r2001_portforward_test.go:6]: done

File: admission/rulebinding/cache/helpers_test.go [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+import (
+	"testing"
+
+	typesv1 "node-agent/pkg/rulebindingmanager/types/v1"
+
+	"github.com/stretchr/testify/assert"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestResourcesToWatch(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (amitschendel) [reply admission/rulebinding/cache/helpers_test.go:14]: done

File: admission/rulebinding/cache/cache_test.go [hunk 19]
```diff
@@ -0,0 +1,30 @@
+package cache
+
+import (
+	"testing"
+
+	"github.com/kubescape/operator/utils"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestNewCache(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (amitschendel) [reply admission/rulebinding/cache/cache_test.go:26]: done

**Review by matthyx — DISMISSED** [review]

**Review by matthyx — APPROVED** [review]

**Pull request merged**